File: branch_boundary.cpp
#include <iostream>
#include <algorithm> // Added for sorting
#include <fstream>
#include <vector>
#include <queue>
#include <chrono> // Added for time measurement
using namespace std;
using namespace std::chrono;

struct Item
{
    int value;
    int weight;
    double density; // value per weight
};

struct Node
{
    int level;
    int value;
    int weight;
    vector<int> solution;
};

bool operator<(const Node &a, const Node &b)
{
    return a.value < b.value;
}

bool compare(Item a, Item b)
{
    return a.density > b.density;
}

int max_value = 0;
vector<int> best_solution;

void branchAndBound(int n,vector<Item>& items, int capacity)
{
    priority_queue<Node> pq;
    sort(items.begin(), items.end(), compare);

    Node root;
    root.level = -1;
    root.value = 0;
    root.weight = 0;
    root.solution = vector<int>(n, 0);
    pq.push(root);

    while (!pq.empty())
    {
        Node current = pq.top();
        pq.pop();

        int level = current.level + 1;
        int value = current.value;
        int weight = current.weight;
        vector<int> solution = current.solution;

        // get a new solution
        if (level == n)
        {
            if (value > max_value)
            {
                max_value = value;
                best_solution = solution;
            }
            continue;
        }

        // Include the item at the current level
        if (weight + items[level].weight <= capacity)
        {
            solution[level] = 1;
            pq.push({level, value + items[level].value, weight + items[level].weight, solution});
        }

        // Exclude the item at the current level, and calculate the remaining maximum value, if it is greater than the current maximum value, then push it to the priority queue
        solution[level] = 0;
        int remaining_capacity = capacity - weight;
        int remaining_max_value = 0;
        for(int i = level+1; i < n; i++)
            if(remaining_capacity >= items[i].weight)
            {
                remaining_capacity -= items[i].weight;
                remaining_max_value += items[i].value;
            }
            else
            {
                remaining_max_value += items[i].density * remaining_capacity;
                break;
            }
        if (value + remaining_max_value > max_value)
            pq.push({level, value, weight, solution});
    }
}

int main()
{
    string filepath = "large_scale\\knapPI_3_200_1000_1";
    ifstream file(filepath);

    if (!file.is_open())
    {
        cout << "Failed to open file." << endl;
        return 0;
    }

    // Read the size and capacity
    int n, c;
    if (!(file >> n >> c))
    {
        cerr << "Failed to read the size and capacity from the file." << endl;
        return 1;
    }

    // Read the values and weights
    vector<Item> items(n);
    for (int i = 0; i < n; i++)
    {
        if (!(file >> items[i].value >> items[i].weight))
        {
            cerr << "Failed to read the values and weights from the file." << endl;
            return 1;
        }
        items[i].density = (double)items[i].value / items[i].weight;
    }

    file.close();

    auto start = high_resolution_clock::now();

    branchAndBound(n,items, c);

    auto end = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(end - start);

    /*cout << "0/1 Sequence: [";
    for (int i = 0; i < n; i++)
    {
        cout << best_solution[i] << ",";
    }
    cout << "]" << endl;*/

    cout << "Maximum Value: " << max_value << endl;

    // Print the execution time
    cout << "Execution Time: " << duration.count() << " microseconds" << endl;

    return 0;
}

File: pycode/branch_boundary.py
## 这个代码山存在问题，需要修改
import heapq
from typing import List

class Item:
    def __init__(self, value: int, weight: int, density: float):
        self.value = value
        self.weight = weight
        self.density = density

class Node:
    def __init__(self, level: int, value: int, weight: int, solution: List[int]):
        self.level = level
        self.value = value
        self.weight = weight
        self.solution = solution

    def __lt__(self, other):
        return self.value < other.value

def compare(a: Item, b: Item) -> bool:
    return a.density > b.density

max_value = 0
def branchAndBound(n: int, items: List[Item], capacity: int):
    max_value = 0
    best_solution = []

    pq = []
    heapq.heapify(pq)

    items.sort(key=compare)

    root = Node(-1, 0, 0, [0] * n)
    heapq.heappush(pq, root)

    while pq:
        current = heapq.heappop(pq)

        level = current.level + 1
        value = current.value
        weight = current.weight
        solution = current.solution.copy()

        if level == n:
            if value > max_value:
                max_value = value
                best_solution = solution
            continue

        if weight + items[level].weight <= capacity:
            solution[level] = 1
            heapq.heappush(pq, Node(level, value + items[level].value, weight + items[level].weight, solution))

        solution[level] = 0
        remaining_capacity = capacity - weight
        remaining_max_value = 0
        for i in range(level + 1, n):
            if remaining_capacity >= items[i].weight:
                remaining_capacity -= items[i].weight
                remaining_max_value += items[i].value
            else:
                remaining_max_value += items[i].density * remaining_capacity
                break

        if value + remaining_max_value > max_value:
            heapq.heappush(pq, Node(level, value, weight, solution))

filepath = "low-dimensional\\f1_l-d_kp_10_269"
with open(filepath, "r") as file:
    lines = file.readlines()

n, c = map(int, lines[2].split())
items = []
for line in lines[4:4+n]:
    value, weight = map(int, line.split())
    density = value / weight
    items.append(Item(value, weight, density))

branchAndBound(n, items, c)

print("Maximum Value:", max_value)
